controller/controller.go

```diff
@@ -451,31 +424,70 @@ func (c *Controller) bootWithError(ctx context.Context) error {
 		}
 	}
 
-	// Initializing the watch gives us all necessary event channels we need to
-	// further process them within the controller. Once we got the event channels
-	// everything is set up for the operator's reconciliation. We put the
-	// controller into a booted state by closing its booted channel so users know
-	// when to go ahead. Note that ProcessEvents below blocks the boot process
-	// until it fails.
+	var mgr manager.Manager
 	{
-		c.logger.LogCtx(ctx, "level", "debug", "message", "starting list-watch")
+		o := manager.Options{
+			// MetricsBindAddress is set to 0 in order to disable it. We do this
+			// ourselves.
+			MetricsBindAddress: "0",
+			SyncPeriod:         to.DurationP(c.resyncPeriod),
+		}
 
-		deleteChan, updateChan, errChan := c.informer.Watch(ctx)
-		close(c.booted)
+		mgr, err = manager.New(c.k8sClient.RESTConfig(), o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var ctrl controller.Controller
+	{
+		o := controller.Options{
+			MaxConcurrentReconciles: 1,
+			Reconciler:              c,
+		}
 
-		c.logger.LogCtx(ctx, "level", "debug", "message", "processing object events")
+		ctrl, err = controller.New(c.name, mgr, o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
 
-		err := c.ProcessEvents(ctx, deleteChan, updateChan, errChan)
+	// Initializing the watch means to have the operator's reconciliation set up.
+	// We put the controller into a booted state by closing its booted channel so
+	// users know when to go ahead. Note that mgr.Start below blocks the boot
+	// process until it ends gracefully or fails.
+	{
+		fmt.Printf("controller starts to watch\n")
+		err = ctrl.Watch(&source.Kind{Type: c.runtimeObjectFactory()}, &handler.EnqueueRequestForObject{})
 		if err != nil {
 			return microerror.Mask(err)
 		}
+		close(c.booted)
+		fmt.Printf("controller signals it has booted\n")
 
-		c.logger.LogCtx(ctx, "level", "debug", "message", "processed object events")
+		err = mgr.Start(setupSignalHandler())
+		if err != nil {
+			return microerror.Mask(err)
+		}
 	}
 
 	return nil
 }
 
+func setupSignalHandler() (stopCh <-chan struct{}) {
```

Comment:
Upstream provides the signal handler, but it panics when you use it multiple times, which we do in our e2e tests. Totally incomprehensible for me why you should ever make this thing panic when being re-used. I shamelessly copied and modified that thing just to make it work. See `mgr.Start` above.

Reply:
My guess would be they want to enforce the design of only ever running one manager within a process. We do this differently ATM.